core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,111 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.resolver.dns.DnsQueryLifecycleObserver;
+
+/**
+ * A {@link DnsQueryLifecycleObserver} that helps capture custom dns metrics.
+ */
+final class DefaultDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
+
+    private final Counter success;
+    private final Counter failure;
+    private final Counter protocolType;
+    private final Counter dnsErrorCode;
+    private final Counter queryWritten;
+    private final Counter queryType;
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry, DnsQuestion question, MeterIdPrefix prefix) {
+        success = meterRegistry.counter(prefix.name(),
+                prefix.withTags("name", question.name(), "result", "success",
+                        "question", question.name()).tags());
+        failure = meterRegistry.counter(prefix.name(),
+                prefix.withTags("name", question.name(), "result", "failure",
+                        "question", question.name()).tags());
+        dnsErrorCode = meterRegistry.counter(prefix.name(),
+                prefix.withTags("name", question.name(), "dns", "errorcodes",
+                        "question", question.name()).tags());
+        queryWritten = meterRegistry.counter(prefix.name(),
+                prefix.withTags("name", question.name(), "written", question.type().name()).tags());
+        queryType = meterRegistry.counter(prefix.name(),
+                prefix.withTags("name", question.name(),
+                        "type", question.type().name()).tags());
+        protocolType = meterRegistry.counter(prefix.name(),
+                prefix.withTags("name", question.name(), "protocol", getProtocolType(question.type()),
+                        "question", question.name()).tags());
+        protocolType.increment();
```

Comment:
There are many incorrectness in this block, so I'm not sure how I can advise on this. Please make sure to re-read my suggestions about the *names* and *tags* and try re-writing this block completely.

For example, you should have different counter for each error code. To do this, you must get the counter lazily in the callback.

Reply:
Regarding this - Your original comment -
All meters with the same name must have the same set of tags in Prometheus, i.e. this will fail. How about organizing like this:
```
TIMER <prefix>{name=<domain name>, result=<success_or_failure>, cause=<cause_of_failure>}
COUNTER <prefix>.written{name=<domain_name>, server=<dns_server_ip_address>}
COUNTER <prefix>.cancelled{name=<domain_name>}
COUNTER <prefix>.cnamed{name=<domain_name>, cname=<name_in_cname_question>}
COUNTER <prefix>.redirected{name=<domain_name>, servers=<list_of_name_servers>}
COUNTER <prefix>.noanswer{name=<domain_name>, code=<response_code>}
```
The prefix here is something default like 'armeria.client.dns.queries'. So <prefix>.cnamed is the name of the Counter ? and everything inside {} are the tags for that counter? Maybe we can pass the metricRegistry to DefaultDnsQueryLifecycleObserver and use it to instantiate the necessary counters when the appropriate dns lifecycle methods are being called?

Reply:
> So .cnamed is the name of the Counter ? 

Yes, its name will be `armeria.client.dns.queries.cnamed` and its tags will be `name=<....>` and `cname=<...>`. The counter will be increased when `queryCNAMEd()` method is invoked.

> everything inside {} are the tags for that counter?

That's correct. :smile: 

> Maybe we can pass the metricRegistry to DefaultDnsQueryLifecycleObserver and use it to instantiate the necessary counters when the appropriate dns lifecycle methods are being called?

Yes! :+1: